### ISSUE TITLE
Value Checker gives up on shifts and masks for ranges

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#1623]
`Range.java` doesn't attempt to determine a range of values for shifts or masks (bitwise ands). Consider the following code (adapted from plume-lib):

```java
import org.checkerframework.common.value.qual.IntRange;

public class ShiftAndMaskExample {
    public static void hexEncode(byte[] bytes) {
        for (int i = 0; i < bytes.length; i++) {
            byte b = bytes[i];
            @IntRange(from = 0, to = 15) int i1 = (b & 0xf0) >> 4;
            @IntRange(from = 0, to = 15) int i2 =  b & 0x0f;
        }
    }
}
```

Though both `@IntRange` annotations are true, the Value Checker isn't able to figure either out because `Range.java` doesn't attempt to handle either the shifts or bitwise and.

**Comment by JasonMrX** [comment]
Hi @kelloggm, 

The `Range.java` library indeed gives up on the calculation of most of the bitwise operations like bitwise and:
``` java
    /** We give up the analysis for bitwise AND operation */
    public Range bitwiseAnd(Range right) {
        return EVERYTHING;
    }
```
The reason is that the scenarios in general are way more complicate than other numeric operations. For example, consider when the mask (the right operand) is of a wide range, it's really difficult for me to come up with a simple refinement strategy that yields a range for bitwise and operation that is precise enough (see the range modulo operation for an example of how I use a relatively simple refinement strategy that provides range refinement results that are loosen but still precise enough).

However, after seeing this sample you provided, I am thinking that maybe we can do a trade-off here -- instead of considering the general cases where the mask could take many possible values, we could only consider the case where the mask is a constant. The problem is greatly simplify in this case, and I think this is the most common scenario in real life codes.

What do you think? Any suggestions are appreciated!

Jason

**Comment by mernst** [comment]
I agree. For this case, it is enough to handle constant shifts, and that will be the most common case in practice.

**Comment by kelloggm** [comment]
Fixed for most practical cases in #1634
